lib/sequel/extensions/connection_metrics.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+#
+# The connection_metrics extension enhances a database's
+# connection pool to store metrics about the use of connections.
+# Whenever a connection is acquired or released the number of
+# currently acquired connections is emitted. Same for making new
+# connections and disconnecting them from the DB. Example of use:
+#
+#   DB.extension(:connection_metrics)
+#
+# Note that this extension has been only tested with a
+# connection pool type :threaded. To use it with other
+# connection pool types it might need adjustments.
+#
+# Related module: Sequel::ConnectionMetrics
+
+module Sequel
+  module ConnectionMetrics
+    # Initialize the data structures used by this extension.
+    def self.extended(pool)
+      case pool.pool_type
+      when :single, :sharded_single, :timed_queue, :sharded_threaded, :sharded_timed_queue
+        raise Error.new('cannot load connection_metrics extension if using a connection pool type different than :threaded')
+      end
+
+      pool.instance_exec do
+        sync do
+          @prometheus_updater = CloudController::DependencyLocator.instance.prometheus_updater
+          @connection_info = {}
+        end
+      end
+    end
+
+    private
+
+    def acquire(thread)
+      begin
+        if (conn = super)
+          acquired_at = Time.now
+          @prometheus_updater.increment_gauge_metric(:cc_acquired_db_connections_total, labels: { process_type: })
+          @connection_info[thread] ||= {}
+          @connection_info[thread][:acquired_at] = acquired_at
+        end
+      rescue Sequel::PoolTimeout
+        @prometheus_updater.increment_gauge_metric(:cc_db_connection_pool_timeouts_total, labels: { process_type: })
+        raise
+      ensure
+        # acquire calls assign_connection, where the thread possibly has to wait for a free connection.
+        # For both cases, when the thread acquires a connection in time, or when it runs into a PoolTimeout,
+        # we emmit the time the thread waited for a connection.
+        if @connection_info[thread] && @connection_info[thread].key?(:waiting_since)
+          @prometheus_updater.update_histogram_metric :cc_db_connection_wait_duration_seconds, (Time.now - @connection_info[thread][:waiting_since]).seconds
+          @connection_info[thread].delete(:waiting_since)
+        end
+      end
+      conn
+    end
+
+    def assign_connection(thread)
+      conn = super
+
+      # if assign_connection does not return a connection, the pool is exhausted and the thread has to wait
+      unless conn
+        waiting_since = Time.now
+        @connection_info[thread] = { waiting_since: } unless @connection_info[thread] && @connection_info[thread].key?(:waiting_since)
+      end
+      conn
+    end
+
+    def make_new(server)
+      conn = super
```

Comment:
In this case we need it, because we would have to put super at the end of the method, to return the connection. But `size` will be updated by super, so we have to emit the metric after super has returned. Additionally if super fails, we'd need error handling. So I think this way it's easier.